android/src/main/java/com/brentvatne/exoplayer/ExoPlayerView.kt

```diff
@@ -0,0 +1,298 @@
+        subtitleLayout.layoutParams = layoutParams
+        subtitleLayout.setUserDefaultStyle()
+        subtitleLayout.setUserDefaultTextSize()
+
+        updateSurfaceView(viewType)
+
+        adOverlayFrameLayout = FrameLayout(context)
+
+        layout.addView(shutterView, 1, layoutParams)
+        layout.addView(adOverlayFrameLayout, 2, layoutParams)
+
+        addViewInLayout(layout, 0, aspectRatioParams)
+        addViewInLayout(subtitleLayout, 1, layoutParams)
+    }
+
+    private fun clearVideoView() {
+        when (surfaceView) {
+            is TextureView -> player?.clearVideoTextureView(surfaceView as TextureView)
+
+            is SurfaceView -> player?.clearVideoSurfaceView(surfaceView as SurfaceView)
+
+            else -> {
+                Log.w("clearVideoView", "Unexpected surfaceView type: ${surfaceView?.javaClass?.name}")
+            }
+        }
+    }
+
+    private fun setVideoView() {
+        when (surfaceView) {
+            is TextureView -> player?.setVideoTextureView(surfaceView as TextureView)
+
+            is SurfaceView -> player?.setVideoSurfaceView(surfaceView as SurfaceView)
+
+            else -> {
+                Log.w("setVideoView", "Unexpected surfaceView type: ${surfaceView?.javaClass?.name}")
+            }
+        }
+    }
+
+    fun setSubtitleStyle(style: SubtitleStyle) {
+        // ensure we reset subtitle style before reapplying it
+        subtitleLayout.setUserDefaultStyle()
+        subtitleLayout.setUserDefaultTextSize()
+
+        if (style.fontSize > 0) {
+            subtitleLayout.setFixedTextSize(TypedValue.COMPLEX_UNIT_SP, style.fontSize.toFloat())
+        }
+        subtitleLayout.setPadding(style.paddingLeft, style.paddingTop, style.paddingTop, style.paddingBottom)
+        if (style.opacity.toInt() != 0) {
+            subtitleLayout.alpha = style.opacity
+            subtitleLayout.visibility = View.VISIBLE
+        } else {
+            subtitleLayout.visibility = View.GONE
+        }
+    }
+
+    fun setShutterColor(color: Int) {
+        shutterView.setBackgroundColor(color)
+    }
+
+    fun updateSurfaceView(@ViewType.ViewType viewType: Int) {
+        this.viewType = viewType
+        var viewNeedRefresh = false
+        when (viewType) {
+            ViewType.VIEW_TYPE_SURFACE, ViewType.VIEW_TYPE_SURFACE_SECURE -> {
+                if (surfaceView !is SurfaceView) {
+                    surfaceView = SurfaceView(context)
+                    viewNeedRefresh = true
+                }
+                (surfaceView as SurfaceView).setSecure(viewType == ViewType.VIEW_TYPE_SURFACE_SECURE)
+            }
+
+            ViewType.VIEW_TYPE_TEXTURE -> {
+                if (surfaceView !is TextureView) {
+                    surfaceView = TextureView(context)
+                    viewNeedRefresh = true
+                }
+                // Support opacity properly:
+                (surfaceView as TextureView).isOpaque = false
+            }
```

Comment:
It would be nice to remove `as` type cast.
I've written an sample, but you can do whatever implementation you're comfortable with.
```kotlin
            ViewType.VIEW_TYPE_SURFACE, ViewType.VIEW_TYPE_SURFACE_SECURE -> {
                surfaceView?.let { it as? SurfaceView }?.apply {
                    setSecure(viewType == ViewType.VIEW_TYPE_SURFACE_SECURE)
                } ?: {
                    surfaceView = SurfaceView(context)
                    viewNeedRefresh = true
                }
            }

            ViewType.VIEW_TYPE_TEXTURE -> {
                surfaceView?.let { it as? TextureView }?.apply {
                    // Support opacity properly:
                    isOpaque = false
                } ?: {
                    surfaceView = TextureView(context)
                    viewNeedRefresh = true
                }
            }
```

Reply:
In the code above, the `surfaceView?.let { it as? SurfaceView }` part is seen as type resolution syntax. 

However, to me, it appears to be a readability breaker. That's because when you write `as?”, people that read codes have to be aware that null is returned if it doesn't match the type on the right, in order to understand the code. I

Reply:
If you think it breaks readability, feel free to ignore above sample :)
I was going to say that any code is fine that can remove `as type cast`.